### ISSUE TITLE
Missed optimization: per-byte comparisons are not always combined properly

### ISSUE THREAD

**Issue by purplesyringa** [llvm/llvm-project#117853]
Reproducer: https://godbolt.org/z/6caGGoo1e

```c
int is_all_ones(unsigned char *p) {
    unsigned char a = p[0];
    unsigned char b = p[1];
    return a == 255 && b == 255;
}
```

I've expected this to compile to a single read + comparison on platforms that support unaligned accesses. Instead, this happened:

```x86asm
is_all_ones:
        movzx   ecx, byte ptr [rdi + 1]
        and     cl, byte ptr [rdi]
        xor     eax, eax
        cmp     cl, -1
        sete    al
        ret
```

**Comment by RKSimon** [comment]
```ll
define range(i32 0, 2) i32 @is_all_ones(ptr nocapture noundef readonly %p) {
entry:
  %0 = load i8, ptr %p, align 1
  %arrayidx1 = getelementptr inbounds i8, ptr %p, i64 1
  %1 = load i8, ptr %arrayidx1, align 1
  %cmp = icmp eq i8 %0, -1
  %cmp3 = icmp eq i8 %1, -1
  %2 = select i1 %cmp, i1 %cmp3, i1 false
  %conv4 = zext i1 %2 to i32
  ret i32 %conv4
}
```

**Comment by PhilippRados** [comment]
@purplesyringa so you mean only requiring a single word-read if both elements are next to each other like this?
```
_is_all_ones:
        movzx  ecx, word ptr [rdi]
        xor    eax, eax
        cmp    cx, -1
        sete   al
        ret
```

**Comment by purplesyringa** [comment]
Yes (obviously, with arbitrary constants, not just `-1`). Ideally,

```c
int is_all_ones(unsigned char *p, unsigned char *q) {
    return (p[0] == q[0]) & (p[1] == q[1]);
}
```

would compile to a single comparison, too.

**Comment by PhilippRados** [comment]
@RKSimon Do you have a suggestion where this should be implemented (should it even)? I saw that there was once a dedicated `LoadCombine` pass which was [removed](https://github.com/llvm/llvm-project/commit/5a7c2f17009460251cc27d9bda183d3272a419ba#diff-cb473bc046b3d5bfb74a81830014f32396161ab4447edc0a20cea7015cf96f3f) though.

According to this [discussion](https://groups.google.com/g/llvm-dev/c/ZwLWNxCY5wE) merging adjacent loads is still being performed (as of 2019) however I couldn't pinpoint where exactly (InstCombine, MemCpyOptimizer, SelectionDAG?).

**Comment by nikic** [comment]
In the middle-end, the place to do it would be AggressiveInstCombine, which already has a transform for merging loads (but not targeting comparisons).

**Comment by nikic** [comment]
We could also view this in terms of forming a memcmp (which later gets reexpanded) via MergeICmps. I think that pass may be more convenient in terms of existing infrastructure to match such patterns (including for the non-trivial cases spread across many blocks).

**Comment by RKSimon** [comment]
FTR, if this had been 4 or more (pow2) comparisons, SLP would have caught it - but it avoids creating reductions of 2 x vectors.
```c
int is_all_ones(unsigned char *p, unsigned char *q) {
    return (p[0] == q[0]) & (p[1] == q[1]) & (p[2] == q[2]) & (p[3] == q[3]);
}
```
